pygsti/algorithms/germselection.py

```diff
@@ -1529,15 +1744,41 @@ def find_germs_breadthfirst(model_list, germs_list, randomize=True,
 
                 for k, model in enumerate(model_list):
                     currentDDDList[k] += _compute_twirled_ddd(
-                        model, germs_list[goodGermIdx], tol)
+                        model, germs_list[goodGermIdx], tol, float_type=float_type)
 
         #aggregate each currendDDDList across all procs
         if comm is not None and comm.Get_size() > 1:
             for k, model in enumerate(model_list):
-                result = _np.empty((Np, Np), 'complex')
+                result = _np.empty((Np, Np), dtype=float_type)
                 comm.Allreduce(currentDDDList[k], result, op=MPI.SUM)
                 currentDDDList[k][:, :] = result[:, :]
                 result = None  # free mem
+                
+    elif mode== "compactEVD":
+        #implement a new caching scheme which takes advantage of the fact that the J^T J matrices are typically
+        #rather sparse. Instead of caching the J^T J matrices for each germ we'll cache the compact SVD of these
+        #and multiply the compact SVD components through each time we need one.
+        twirledDerivDaggerDerivList = \
+            [_compute_bulk_twirled_ddd_compact(model, germs_list, tol,
+                                              evd_tol=1e-10, float_type=float_type, printer=printer)
+             for model in model_list]
+             
+             #_compute_bulk_twirled_ddd_compact returns a tuple with three lists
+             #corresponding to the u, sigma and vh matrices for each germ's J^T J matrix's_list
+             #compact svd.
+        currentDDDList = []
+        nonzero_weight_indices= _np.nonzero(weights)
+        nonzero_weight_indices= nonzero_weight_indices[0]
+        for i, derivDaggerDeriv in enumerate(twirledDerivDaggerDerivList):
+            #reconstruct the needed J^T J matrices
+            for j, idx in enumerate(nonzero_weight_indices):
+                if j==0:
+                    temp_DDD = derivDaggerDeriv[0][idx] @ derivDaggerDeriv[2][idx]
+                else:
+                    temp_DDD += derivDaggerDeriv[0][idx] @ derivDaggerDeriv[2][idx]
+                    
+                #print('temp_DDD shape= ',temp_DDD.shape) 
+            currentDDDList.append(temp_DDD)
 
     else:  # should be unreachable since we set 'mode' internally above
```

Comment:
This is no longer unreachable - should remove this comment.

Reply:
Done